---
title: "Menu"
output: 
  github_document:
    toc: true
    toc_depth: 4
editor_options: 
  markdown: 
    wrap: 72
---

<!-- README.md is generated from README.Rmd. Please edit that file -->

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  fig.path = "man/figures/README-",
  out.width = "100%"
)
```

# SCSES

<!-- badges: start -->

<!-- badges: end -->

Single-cell Splicing Estimation based on Network Diffusion

## Hardware requirements

`SCSES` package requires only a standard computer with enough RAM to
support the in-memory operations.

## Software requirements

### OS Requirements

This package is supported for Linux.

The package has been tested on Debian-11.21.

## Installation

### Installation with docker file

SCSES provides a Docker-based installation method to simplify the setup
of all dependencies and requirements. Please follow the steps below to
build the Docker image and start a container to use SCSES:

#### 1. Install Docker Client:

Please install the [Docker
client](https://www.docker.com/products/docker-desktop) on the host
machine.

#### 2. Download Dockerfile

The Dockerfile of SCSES can be downloaded from:
<https://github.com/lvxuan12/SCSES/blob/main/SCSES.dockerfile>.

#### 3. Build Docker Image

You can build the SCSES Docker image using the command:

``` bash
docker build -t scses -f SCSES.dockerfile .
```

#### 4. Create Docker Container

After building the image, create a Docker container with the following
command:

``` bash
docker run -d -p [exported port]:8787 -e PASSWORD=[user password] -v [local directory]:/data --name test scses
```

`[exported port]`: The port on the host machine to access the container.

`[user password]`: A user-defined password for logging into the RStudio
server.

`[local directory]`: A local directory mapped to the container for data
storage and sharing.

#### 5. Access RStudio Server

Now, you can access the RStudio server by opening a web browser and
navigating to `[host IP]:[exported port]`. Use the default username
`rstudio` and the user-defined password to log in.

In this pre-configd RStudio server environment, SCSES and all its
dependencies are correctly installed and ready for use.

Please refer to [Getting started](#getting-started) to start the first
experience with SCSES.

Enjoy!

### Installation with conda

#### Step 1: Environment Setup

We recommend a new **conda** environment to install SCSES:

``` bash
conda create -n SCSES_test python=3.11
conda activate SCSES_test
```

To use SCSES, you will need to install R, Python, Matlab Compiler
Runtime(v9.13), and Java(v17.0.10).

``` bash
## install R in conda environment
conda install -c conda-forge r-base=4.3.1
```

The MCR is quite large, so downloading may take some time.

``` bash
## install MCR
mkdir /path/to/MCR && \
cd /path/to/MCR && \
wget https://ssd.mathworks.com/supportfiles/downloads/R2022b/Release/10/deployment_files/installer/complete/glnxa64/MATLAB_Runtime_R2022b_Update_10_glnxa64.zip && \
unzip -q MATLAB_Runtime_R2022b_Update_10_glnxa64.zip && \
./install -destinationFolder /opt/mcr -agreeToLicense yes -mode silent
```

#### Step 2: Install Python Dependencies

``` bash
pip install pandas numpy scipy scikit-learn
pip install keras==2.15.0
pip install tensorflow==2.15.0.post1
```

#### Step 3: Install Tools

To detect splicing events You will need to install rMATS, MAJIQ,
IRFinder. rMATS should be built in the same environment with SCSES (same
python). MAJIQ should be built in a new environment due to conflict of
python package version，

##### 3.1 [rMATS](https://github.com/Xinglab/rmats-turbo)

``` bash
wget https://github.com/Xinglab/rmats-turbo/releases/download/v4.3.0/rmats_turbo_v4_3_0.tar.gz
tar -zxvf rmats_turbo_v4_3_0.tar.gz
cd rmats_turbo_v4_3_0/
# Install dependencies
pip install Cython
# Add to PATH
./build_rmats
export PATH=/path/to/rmats_turbo_v4_3_0/:$PATH
```

##### 3.2 [MAJIQ](https://biociphers.bitbucket.io/majiq-docs/index.html)

``` bash
conda create -n MAJIQ python=3.11
conda activate MAJIQ

# install HTSlib from <https://www.htslib.org/download/>
# change to where library/include directories are installed

export HTSLIB_LIBRARY_DIR=/path/to/htslib-1.15.1/lib/
export HTSLIB_INCLUDE_DIR=/path/to/htslib-1.15.1/include/

pip install git+https://bitbucket.org/biociphers/majiq_academic.git
export MAJIQ_LICENSE_FILE=/path/to/majiq_license_academic_official.lic
```

**NOTE:** MAJIQ will not function without providing the license file.

##### 3.3 [IRFinder](https://github.com/dgaolab/IRFinder)

``` bash
wget https://github.com/RitchieLabIGH/IRFinder/archive/refs/tags/v2.0.1.tar.gz
tar -zxvf v2.0.1.tar.gz
export PATH=/path/to/IRFinder-2.0.1/bin/:$PATH
```

**NOTE:** [STAR](https://github.com/alexdobin/STAR) is required to build
IRFinder reference! To run IRFinder correctly, you also need to install
STAR.

##### 3.4 [samtools](https://github.com/samtools/samtools)

#### Step 4: Install SCSES R Package

Currently SCSES can only be installed from GitHub.

To install SCSES, type the following command in **R**:

``` r
install.packages("remotes")
remotes::install_version("Matrix", version = "1.6-5")
install.packages("curl",config.vars='LIB_DIR=/usr/lib/x86_64-linux-gnu/pkgconfig/')
options(download.file.method = "wget", times=100)
remotes::install_github("lvxuan12/SCSES")
```

#### Tips for some Installation errors

##### Issue 1. cannot find fftw.h

``` bash
conda install conda-forge::fftw
# config FFTW
export FFTW_CFLAGS=" -I/path/to/miniconda3/envs/scses/include/"
export FFTW_LIBS=" -L/path/to/miniconda3/envs/scses/lib -lfftw3"
```

##### Issue 2. cannot find -lxml2

``` bash
conda install conda-forge::libxml2
```

##### Issue 3. cannot find -lsz

``` bash
ln -s /usr/lib/x86_64-linux-gnu/libsz.so /path/to/miniconda/envs/SCSES_test/lib/libsz.so
```

##### Issue 4. error: '::timespec_get' has not been declared

``` bash
conda upgrade -c conda-forge --all 
```

## SCSES input

SCSES requires five essential input files:

### 1. BAM Files

-   **Format**: Coordinate-sorted BAM files with index (.bai)

### 2. Reference Genome Files

-   **FASTA**: Reference genome sequence
-   **GTF**: Gene annotations
-   **GFF3**: Gene annotations

### 3. Configuration File

A config file is required to run SCSES.

You can use `createConfigshiny` command to generate a config file:

If you used the Docker-based installation method:

``` r
library(SCSES)
createConfigshiny(host = "localhost",launch.browser=TRUE) 
```

For command line environment, it is recommended to:

Setting launch.browser = TRUE may cause errors in headless environments
(servers without GUI) or when no default browser is configd

``` r
library(SCSES)
createConfigshiny(host, port, launch.browser=FALSE) 
```

#### 3.1. Set the host to the server's IP address

#### 3.2. Set launch.browser = FALSE to avoid browser launch errors

#### 3.3. Manually access the application URL shown in the console

After running `createConfigshiny`, you will see a URL appear in the
console

Copy this URL and paste it into your web browser to access the
application.

After that, a interactive window will popup which allow you to fill some
parameters, such as Bam File Path, and Work Path.

Finally, you can click “Create Config” button and a json file will be
generated in the `work_path` you provided if successful.

For a detailed explanation of the configuration file, please refer to
the
[ConfigurationGuide.txt](https://github.com/lvxuan12/SCSES/blob/main/ConfigurationGuide.txt).

### 4. Phast conservation file in bigWig format

For human and mouse, you could download it directly from UCSC browser:

| Species      | File                     | Size   | Link                                                                                                |
|-------------------|------------------|------------------|------------------|
| Human (hg38) | hg38.phastCons100way.bw  | 5.5 GB | [Download](http://hgdownload.cse.ucsc.edu/goldenPath/hg38/phastCons100way/hg38.phastCons100way.bw)  |
| Human (hg19) | hg19.100way.phastCons.bw | 5.4 GB | [Download](http://hgdownload.cse.ucsc.edu/goldenPath/hg19/phastCons100way/hg19.100way.phastCons.bw) |
| Mouse (mm10) | mm10.60way.phastCons.bw  | 4.3 GB | [Download](http://hgdownload.cse.ucsc.edu/goldenPath/mm10/phastCons60way/mm10.60way.phastCons.bw)   |

### 5. RBP

Genes annotated as RBP are required to constructs similarity networks.
For human and mouse, you could read from package:

```{r}
# Load human RBP list
rbp_human <- system.file("extdata/rbp/human_rbp.txt", package = "SCSES")
rbp_list <- readLines(rbp_human)
cat("Total RBPs:", length(rbp_list), "\n")
cat("First 10 RBPs:", head(rbp_list, 10), sep = "\n")
# Load mouse RBP list
rbp_mouse <- system.file("extdata/rbp/mouse_rbp.txt", package = "SCSES")
rbp_list <- readLines(rbp_mouse)
cat("Total RBPs:", length(rbp_list), "\n")
cat("First 10 RBPs:", head(rbp_list, 10), sep = "\n")
```

## Getting started {#getting-started}

### Download Test Data

This dataset includes BAM files for three cell lines (HCT116, HepG2,
HL-60), each containing five cells and other input files that are
essential for running the SCSES package.

**Download**: <https://doi.org/10.5281/zenodo.15688700>

| File Type      | File Name           | Description                                                |
|------------------|---------------------------|---------------------------|
| **BAM**        | `*.bam`             | BAM files for three cell lines, each containing five cells |
| **BAI**        | `*.bam.bai`         | BAM index files                                            |
| **TXT**        | `annotation.txt`    | Cell identities                                            |
| **FASTA**      | `test.fa`           | Reference genome sequence                                  |
| **FAI**        | `test.fa.fai`       | Reference genome sequence index                            |
| **Annotation** | `test.gtf`          | Gene annotation file                                       |
| **Annotation** | `test.gff3`         | Gene annotation file                                       |
| **TXT**        | `human_rbp.txt`     | RNA-binding proteins list                                  |
| **PhastCons**  | `test_phastCons.bw` | Conservation scores                                        |
| **JSON**       | `cell_line.json`    | Parameters config file                                     |

#### Setup

After download, ensure you have:

-   15 BAM files + index files

-   annotation.txt: cell identities

-   test.fa and test.fai: reference genome sequence

-   test.gtf and test.gff: gene annotation file

-   test_phastCons.bw: conservation scores

-   human_rbp.txt: RBP genes list

-   cell_line.json: parameters config file

Move 15 BAM files and their index to `bam` directory.

Move other input data to `refgenome` directory.

```{bash}
# Example:
ls /disk/share/lvxuan/SCSES_test/bam/
ls /disk/share/lvxuan/SCSES_test/refgenome/
```

### Step-by-Step Analysis

#### Step 1. Read config file

The `cell_line.json` file was downloaded previously

```{r Read config file}
## Loading packages
library(SCSES)

# Load configuration file
config_file <- system.file("analysis/cell_line.json", package = "SCSES")
paras <- readSCSESconfig(config_file)

# Verify configuration
cat("Dataset:", paras$DataSet, "\n")
cat("BAM path:", paras$Basic$bam_path, "\n")
# all outputs will be saved in work_path
cat("Work path:", paras$Basic$work_path, "\n")

```

#### Step 2. Get gene expression

##### TPM matrix (for smart-seq2 dataset)

The TPM matrix of gene expression can be obtained by different methods.
We used
[featureCounts](https://github.com/ShiLab-Bioinformatics/subread) to
generate this matrix.

You can use `getGeneExpression` to run `featureCounts`, which will save
featureCounts output to `work_path/expr/` and `getEXPmatrix` to generate
TPM matrix, which will save gene expression count and TPM matrix to
`work_path/rds/`.

```{r Get gene expression}
cat("Quantifying gene expression...\n")
featurecounts.path = getGeneExpression(paras) 
rds.path = getEXPmatrix(paras)

# Load and examine results
print(rds.path)
list.files(rds.path)
tpm = readRDS(paste0(rds.path,'/count_norm.rds'))
tpm[1:5,1:5]
```

After this step, directories `expr` and `rds` will be created.

##### Normalized UMI count matrix (for UMI based dataset)

You can use `get10XEXPmatrix` to generate Normalized UMI count matrix
from 10X CellRanger hdf5 file, which will save normalized UMI count to
`work_path/rds/`.

``` r
# Seurat package is required

rds.path <- get10XEXPmatrix(paras, expr_path)
```

#### Step 3. Detect splicing events

To define a global set of all splicing events, SCSES firstly merges all
bam files from every single cell to construct a pseudo-bulk bam file,
and identifies all types of splicing events by conventional algorithms.

##### For Smart-seq2 dataset

```{r Detect splicing events}
# Create pseudobulk for event detection

pseudobulk.path = createPseudobulk(paras)
print(pseudobulk.path)
list.files(pseudobulk.path)

# Detect all AS events
#if you meet:
#irfinder: error while loading shared libraries: libboost_iostreams.so.1.71.0: cannot open shared object file: No such file or directory
#libboost_iostreams.so.1.71.0 exists in /disk/lvxuan/lib
old.ld=Sys.getenv("LD_LIBRARY_PATH")
Sys.setenv(LD_LIBRARY_PATH = paste0("/disk/lvxuan/lib:", old.ld))

event.path = detectEvents(paras)

# Summary of detected events
print(event.path)
event_files <- list.files(event.path, pattern = "*.txt")
for(file in event_files) {
  events <- readLines(file.path(event.path, file))
  cat(gsub(".txt", "", file), "events:", length(events), "\n")
}

```

Different types of splicing events will be saved to `work_path/events/`,
separately.

##### for UMI dataset

SCSES requires single cell bam files being saved in a directory. For
UMI-based dataset using CellRanger for data process, the function
`split10XBAM` can be used to get single cell bam files for each sample.

``` r
# CellRanger_path: directory to CellRanger output
# out_path: directory to save single cell bam
# core: the number of threads

splitbam.path = split10XBAM(CellRanger_path,out_path,core)

# path to single cell bam files should be added to bam_path in config file
paras$Basic$bam_path = splitbam.path
# pseudobulk.path = createPseudobulk(paras)
# It is not necessary to execute `createPseudobulk`, and the `possorted_genome_bam.bam`,`possorted_genome_bam.bam.bai` from `CellRanger_path` can be moved to `work_path/data/all.bam`.
event.path = detectEvents(paras)
```

#### Step 4. Quantify splicing events

According to splicing events detected in the previous step, SCSES then
quantify raw reads associated with these splicing events in each cell,
and construct the raw read count matrix and calculate the raw PSI
matrix.

The definition of PSI of different AS events: ![The definition of PSI of
different AS events:](png/PSI.png)

```{r Quantify splicing events}
# Raw read counts and PSI calculation
rawrc.path = getRawRC(paras)
rawpsi.path = getRawPSI(paras)
# Merge and preprocess data
rawrds.path = mergeSplicingValue(paras)
processed.data.path = preprocessEvent(paras)
print(rawrds.path)
print(processed.data.path)

# Examine processed data
psi_processed <- readRDS(file.path(processed.data.path, "psi.rds"))
rc_processed <- readRDS(file.path(processed.data.path, "rc.rds"))

cat("Processed PSI matrix:", dim(psi_processed), "\n")
cat("Processed RC matrix:", dim(rc_processed), "\n")

```

Raw read count matrix, PSI matrix, and event annotation will be saved to
`work_path/rds/`.

Then, data after quality control process will be saved to
`work_path/rds_processed/`, which will be used for **subsequent
calculations**.

#### Step 5. Constructs similarity networks

To overcome the high dropout rate and limited read coverage of scRNA-seq
techniques, SCSES constructs cell similarity and event similarity
networks by K-nearest neighbor algorithm (KNN) to learn information from
similar cells and events.

```{r Constructs similarity networks}
# Cell similarity networks
cellnet.path = getCellSimilarity(paras)
# Event similarity networks  
eventnet.path = getEventSimilarity(paras)

# Examine network structure
print(cellnet.path)
cell_networks <- readRDS(file.path(cellnet.path, "cell.similars.rds"))
event_networks <- readRDS(file.path(eventnet.path, "event.similars.rds"))
## three different features can be chosen to quantify cell similarity, including raw event PSI(PSI), and raw junction read counts(RC), and RBP expression(EXP_RBP)
cat("Cell similarity types:", names(cell_networks), "\n")
## different splicing event types
cat("Event types:", names(event_networks), "\n")

```

A list of event similarity for different types of splicing events will
be saved to `work_path/imputation/event_similarity/event.similars.rds`;

A list of different types of cell similarity and a list of the number of
neighbors will be saved to rds file
to`work_path/imputation/cell_similarity/cell.similars.rds` and
`work_path/imputation/cell_similarity/dyk.cell.rds`

In this step, some **parameters** can be adjusted in config file or the
function parameters directly:

For **cell similarity networks**, SCSES can use RBP expressions, Raw
read count or Raw PSI to measure cell similarities.

| Parameter              | Default     | Description                              |
|--------------------|------------------|----------------------------------|
| `feature_num`          | 1000        | Number of high-variable features for PCA |
| `cell_similarity_data` | "EXP_RBP"   | Data types for cell similarity           |
| `distance_method`      | "euclidean" | Distance metric                          |
| `alpha_cell`           | 0.8         | 1 - Random walk restart probability      |
| `kcell_min`            | 5           | Minimum cell neighbors                   |
| `kcell_max`            | 50          | Maximum cell neighbors                   |
| `decay_cell`           | 0.05        | Convergence threshold                    |

For **event similarity networks**, Event similarities are defined by the
RBP regulatory correlations and an embedding representation by
integrating event sequence similarities.

| Parameter     | Default | Description                         |
|---------------|---------|-------------------------------------|
| `kevent`      | 10      | Number of event neighbors           |
| `alpha_event` | 0.8     | 1 - Random walk restart probability |
| `decay_event` | 0.05    | Convergence threshold               |

#### Step 6. Imputation

Based on these weighted similarity networks, SCSES next will use three
imputation strategies to aggregate the information across similar cells
or events to impute read count or PSI value

```{r Imputation}
# Three-strategy imputation
Imputed.data.path = ImputationAll(paras)

# Examine imputation results
Imputed_seperated = readRDS(Imputed.data.path)
str(Imputed_seperated,max.level=3)
# explain:
# For example:
# cell/PSI_PSI: PSI value is used to quantify cell-cell splicing similarity. Impute raw PSI with cell similarities.
# cell/PSI_PSI: PSI value is used to quantify cell-cell splicing similarity. Impute raw inclusion and exclusion read counts with cell similarities, and then calculate the imputed PSI.
# cell_event/PSI_PSI: PSI value is used to quantify cell-cell splicing similarity. Impute raw inclusion and exclusion read counts with cell similarities, and then calculate the imputed PSI. Further impute the results using event similarities.
```

Results of each imputation strategy will be saved
into`work_path/imputation`.

#### Step 7. Estimation

##### 7.1. Estimation based on pre-trained model

We recommend different imputation strategies for four scenarios defined
by the abundance of reads counts in the target cell and neighbor cells
(ND, BD, TD+Info, and TD-Info). SCSES pre-trains models to predict the
probability of specific scenario for each cell-event pair. Finally,
SCSES calculates the PSI value using a linear combination of predictions
from the four strategies, weighted by these probabilities.

```{r Estimation based on pre-trained model}
# Final estimation with scenario classification
# rds_imputed_file: path to the list of three imputation strategies results generated in the previous step
Imputed.data.final.path = Estimation(paras,rds_imputed_file = Imputed.data.path)

# Examine final results
Imputed_combined = readRDS(Imputed.data.final.path)
str(Imputed_combined,max.level=2)
# The finally imputed PSI were named by cell-cell splicing similarity features, including raw event PSI(PSI), and raw junction read counts(RC), and RBP expression(EXP_RBP).
```

A list of final imputation of PSI values will be saved by serialized R
object format (.rds) in `work_path/imputation/Imputed_combined*`, where
**\*** is a random number representing different execution. The `.rds`
file can be loaded in R environment by `readRDS` function.

##### 7.2. Estimation based on fine-tuned model

To improve the fitness of models for a new dataset, we also provide a
procedure to fine-tune the model. For this analysis, we first build a
reference using a set of splicing events with conserved splicing levels
in different human tissues (<https://zenodo.org/records/6408906>). Then
we compare the splicing level in a new dataset with the reference
records, and give the scenarios definition to each event-cell pair,
which is used to fine-tune the pre-trained model.

The commands to perform these analyses:

```{r Estimation based on fine-tuned model}
# Final estimation with scenario classification
ftrc.path = getFtRawRC(paras)
ftpsi.path = getFtRawPSI(paras)
ftrds.path = mergeFtSplicingValue(paras)
ftmodel.path = FtClassifier(paras)
#rds_imputed_file: path to the list of three imputation strategies results generated in the previous step
ImputedFt.data.final.path = Estimation(paras,rds_imputed_file = Imputed.data.path)

print(ImputedFt.data.final.path)

# Examine final results
Imputed_combined = readRDS(ImputedFt.data.final.path)
str(Imputed_combined,max.level=2)
Imputed_combined[["EXP_RBP"]][1:3,1:3]
```

#### Step 8. Cell Clustering

Here is an example of UMAP visualization based on test data:

``` r
install.packages("umap")
```

```{r Cell Clustering}
library(umap)
library(ggplot2)
calcu_umap<-function(data,n_neighbors){
  set.seed(12345)
  data = data[which(apply(data, 1, mean) != 0), ]
  data = data[which(apply(data, 1, var) != 0), ]
  D_Reduct_res <- prcomp(t(data), center = T, scale. = T)
  sdev <- D_Reduct_res$sdev
  var_prop <- sdev^2 / sum(sdev^2)
  cumulative_variance <- cumsum(var_prop)
  n_components <- which(cumulative_variance >= 0.7)[1]
  D_Reduct <- D_Reduct_res$x[, 1:n_components]
  umap_res <- umap::umap(D_Reduct,n_neighbors = n_neighbors)
  input <- umap_res$layout
  df<-as.data.frame(input)
  row.names(df)<-colnames(data)
  return(df)
}

mycol=c("#fbb45d","#699ed4","#ef8183")

# Load data and annotations

if(!file.exists(paste0(system.file("analysis", package = "SCSES"),'/annotation.txt'))) {
  stop("Annotation file not found")
}

if(!file.exists(ImputedFt.data.final.path)) {
  stop("Imputed data file not found")
}

annotation=read.table(paste0(system.file("analysis", package = "SCSES"),'/annotation.txt'),sep="\t")
# Imputed_combined=readRDS(paste0(system.file("analysis", package = "SCSES"),'/Imputed_combined_499997048.rds'))
Imputed_combined=readRDS(ImputedFt.data.final.path)

data_umap=calcu_umap(Imputed_combined[[1]],n_neighbors = 3)
row.names(data_umap)=gsub(".bam","",row.names(data_umap))
data_umap$group=annotation$V2[match(row.names(data_umap),annotation$V1)]

p=ggplot(data = data_umap,aes(x =V1 ,y =V2))+
  geom_point(aes(fill=group),shape=21,size=1.5,stroke=0.05)+
  scale_fill_manual(values = mycol)+
  xlab("UMAP1")+
  ylab("UMAP2")

print(p)

```
