---
title: "SCSES iPSC Tutorial"
output: html_document
date: "2024-10-12"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# SCSES iPSC Tutorial

## Installation

If you havenâ€™t yet installed SCSES, you can find installation instructions in our [GitHub README](https://github.com/lvxuan12/SCSES?tab=readme-ov-file#installation).

## Loading packages

```{r}
library(SCSES)
```

## Loading data

In this tutorial, we will analyse induced human pluripotent stem cells (iPSC), as described in Song et al., 2017. You can run this tutorial by downloading bam files at .

## Prepare data

``` bash
# download phast conservation file
wget http://hgdownload.cse.ucsc.edu/goldenPath/hg19/phastCons100way/hg19.100way.phastCons.bw
```

```{r}
# generate configure file
createConfigshiny(host="127.0.0.1",port = 9999,launch.browser=T) 
```

#### Step1. Read configure file

```{r}
library(SCSES)
#paras_file: path to configure file generated in the previous step
paras_file = "/disk/lvxuan/Single-Splicing/src/package/SCSES/analysis/nPSC.json"
paras = readSCSESconfig(paras_file)
```

#### Step2. Get gene expression

##### TPM matrix

```{r}
featurecounts.path = getGeneExpression(paras) 
rds.path = getEXPmatrix(paras)
```

#### Step3. Detect splicing events

```{r}
pseudobulk.path = createPseudobulk(paras)
#if you meet:
#irfinder: error while loading shared libraries: libboost_iostreams.so.1.71.0: cannot open shared object file: No such file or directory
old.ld=Sys.getenv("LD_LIBRARY_PATH")
Sys.setenv(LD_LIBRARY_PATH = paste0("/disk/lvxuan/lib:", old.ld))
event.path = detectEvents(paras)
```

#### Step4. Quantify splicing events

```{r}
rawrc.path = getRawRC(paras)
rawpsi.path = getRawPSI(paras)
rawrds.path = mergeSplicingValue(paras)
processed.data.path = preprocessEvent(paras)
```

#### Step5. Constructs similarity networks

```{r}
cellnet.path = getCellSimilarity(paras)
eventnet.path = getEventSimilarity(paras)
```

#### Step6. Imputation

```{r}
Imputed.data.path = ImputationAll(paras)
```

A list of three imputation strategies result will be saved to `work_path/imputation/`.

#### Step7. Estimation

```{r}
#rds_imputed_file: path to the list of three imputation strategies results generated in the previous step
Imputed.data.path = Estimation(paras,rds_imputed_file="/disk/lvxuan/Single-Splicing/result/GSE85908/scses_test/imputation/Imputed_seperated_500007276.rds")
```

##### Fine-tune the model

```{r}
ftrc.path = getFtRawRC(paras)
ftpsi.path = getFtRawPSI(paras)
ftrds.path = mergeFtSplicingValue(paras)
ftmodel.path = FtClassifier(paras)
Imputed.data.path = Estimation(paras,rds_imputed_file="/disk/lvxuan/Single-Splicing/result/GSE85908/scses_test/imputation/Imputed_seperated_500007276.rds")
```

A list of final imputation of PSI values will be saved to `work_path/imputation/`.

## Including Plots

You can also embed plots, for example:

```{r pressure, echo=FALSE}
plot(pressure)
```

Note that the `echo = FALSE` parameter was added to the code chunk to prevent printing of the R code that generated the plot.
